### PR TITLE
Some more changes and cleanup to reduce the startup time

### PR DESCRIPTION
**Pull request by daxian-dbw**
## PR Summary

This PR includes some changes and cleanups that will reduce the startup time. Each commit is self-contained, and the commit message serves as a summary of the changes in the commit.
- Rename `s_wasSystemPolicyDebugPolicy` to `s_allowDebugOverridePolicy` to make it less confusing. Also slightly refactor `HelperSecurity.psm1` and `ConstrainedLanguageDebugger.Tests.ps1` to remove unneeded code. There is no functional change in this commit.
- Remove the unneeded static property `IsInbox`, as PowerShell Core won't be shipped in-box with Windows in the foreseeable feature. Even if we do in future, we won't be needing it because Windows PowerShell will probably be gone by that time.
   - This static property was hit at startup time when constructing the value for `$env:PSModulePath`, and it will in turn cause `IsNanoServer` and `IsIoT` to be evaluated, all of which will trigger access to the Registry. By removing `IsInbox`, that can be all avoided.
- Update 'BindRunspace' to avoid getting all commands and unneeded method calls.
   - `BindRunspace` retrieves all available commands from the session state, but for the most common scenario of creating/opening a Runspace, the retrieved commands are never get used afterwards. The method is updated to avoid retrieving all commands unless it's necessary.
   - The method is also updated to add check before calling into a `ProcessXXX` method, so that we can avoid some unneeded method calls.
- Avoid creating a `IsSafeValueVisitor` every time when `IsScriptBlockInFactASafeHashtable` runs.
   - The static method `IsSafeValueVisitor.IsAstSafe` creates an instance of `IsSafeValueVisitor` every time it runs. Given that `IsScriptBlockInFactASafeHashtable` gets called relatively frequently in `PerformSecurityCheck`, this will result in some GC pressure as it will generate transient objects. This is updated to reuse a `IsSafeValueVisitor` singleton with the default `SafeValueContext`.
   - The `Default` and `_safeValueContext` fields are made `readonly`, so it cannot be changed by reflection.
   - The `_visitCount` is changed to `uint` type, so an attacker cannot change the value of it to a negative number in order to practically increase the max visit count limit.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by iSazonov** [comment]
> The _visitCount is changed to uint type, so an attacker cannot change the value of it to a negative number in order to practically increase the max visit count limit.

Currently runtime has `Unsafe.As` method and similar others. Can these methods be used to change the value with reflection?

**Review by iSazonov — APPROVED** [review]

File: src/System.Management.Automation/security/wldpNativeMethods.cs [hunk 1]
```diff
@@ -64,7 +64,7 @@ public static SystemEnforcementMode GetSystemLockdownPolicy()
 
         private static object s_systemLockdownPolicyLock = new Object();
         private static Nullable<SystemEnforcementMode> s_systemLockdownPolicy = null;
-        private static bool s_wasSystemPolicyDebugPolicy = false;
+        private static bool s_allowDebugOverridePolicy = false;
```
Comment (PaulHigin) [inline src/System.Management.Automation/security/wldpNativeMethods.cs]: I feel that this name is no less confusing.  This indicates that the system wide debug policy is in effect.  I feel a better name would be something like: `s_isDebugSystemPolicy`

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 2]
```diff
@@ -2360,67 +2360,104 @@ private void Bind_LoadAssemblies(ExecutionContext context)
...
-            // in his module path, then that will get loaded instead of the actual nested module (from the GAC - in our case)
-            // Hence, searching only from the system module path while loading the core modules
             var unresolvedCmdsToExpose = new HashSet<string>(this.UnresolvedCommandsToExpose, StringComparer.OrdinalIgnoreCase);
-            ProcessImportModule(initializedRunspace, CoreModulesToImport, ModuleIntrinsics.GetPSHomeModulePath(), publicCommands, unresolvedCmdsToExpose);
+            if (CoreModulesToImport.Count > 0 || unresolvedCmdsToExpose.Count > 0)
+            {
+                // If a user has any module with the same name as that of the core module( or nested module inside the core module)
+                // in his module path, then that will get loaded instead of the actual nested module (from the GAC - in our case)
+                // Hence, searching only from the system module path while loading the core modules
+                ProcessImportModule(initializedRunspace, CoreModulesToImport, ModuleIntrinsics.GetPSHomeModulePath(), GetPublicCommands(), unresolvedCmdsToExpose);
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: It looks like GetPublicCommands() result is only used in the `ProcessImportModule` overloaded method, so we don't need to pass it in as a parameter and can just add it there.  Also can we re-name this overload method?  I find it very confusing to have two methods that do different work named the same (and take the same number of parameter arguments).

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/InitialSessionState.cs]: > It looks like GetPublicCommands() result is only used in the ProcessImportModule overloaded method, so we don't need to pass it in as a parameter and can just add it there

If we retrieve the public commands within `ProcessImportModule`, then in the worst case where we do have to call `ProcessImportModule` 3 times in `BindRunspace`, we will have to retrieve the public commands 3 time. That's redundant and should be avoided.

> Also can we re-name this overload method? I find it very confusing to have two methods that do different work named the same (and take the same number of parameter arguments).

I agree it's a bit confusing, especially one overload is only used in the other overload. I will try rename them.

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 3]
```diff
@@ -2360,67 +2360,104 @@ private void Bind_LoadAssemblies(ExecutionContext context)
...
-            // Get initial list of public commands in session.
-            HashSet<CommandInfo> publicCommands = new HashSet<CommandInfo>();
-            foreach (CommandInfo sessionCommand in initializedRunspace.ExecutionContext.SessionState.InvokeCommand.GetCommands(
-                        "*", CommandTypes.Alias | CommandTypes.Function | CommandTypes.Filter | CommandTypes.Cmdlet, true))
-            {
-                if (sessionCommand.Visibility == SessionStateEntryVisibility.Public)
+            // Get initial list of public commands from session in a lazy way.
+            // We could use Lazy<> with an initializer for the same purpose, but we can save allocations
+            // by using the local function. It avoids allocating the delegate, and it's more efficient on
+            // capturing variables from the enclosing scope by using a struct.
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/InitialSessionState.cs]: It's not only for startup time, but also for creating/opening a Runspace in common cases. I will update the comment a bit.

File: src/System.Management.Automation/security/wldpNativeMethods.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/security/wldpNativeMethods.cs]: `s_wasSystemPolicyDebugPolicy` will be set to `true` as long as we can reach the `GetDebugLockdownPolicy(string)` method, and it doesn't matter whether or not the debug policy is set and enforced.

That means, as long as the system lock-down policy is not `Enforce`, it's allowed to use debug override policy, and thus we will check for the debug policy every time `GetSystemLockdownPolicy()` is called.

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by daxian-dbw** [comment]
> Currently runtime has Unsafe.As method and similar others. Can these methods be used to change the value with reflection?

@iSazonov I don't think `uint` field can be set with a negative value by reflection.

**Pull request merged**